internal/core/metadata/operators/device_service/get_test.go

```diff
@@ -0,0 +1,70 @@
+/*******************************************************************************
+ * Copyright 2019 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package device_service
+
+import (
+	"testing"
+
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/operators/device_service/mocks"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/config"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+)
+
+func TestGetAllDeviceServices(t *testing.T) {
+	tests := []struct {
+		name        string
+		cfg         config.ServiceInfo
+		dbMock      DeviceServiceLoader
+		expectError bool
+	}{
+		{"GetAllPass", config.ServiceInfo{MaxResultCount: 1}, createGetDeviceServiceLoaderMock(), false},
+		{"GetAllFailCount", config.ServiceInfo{}, createGetDeviceServiceLoaderMock(), true},
+		{"GetAllFailUnexpected", config.ServiceInfo{MaxResultCount: 1}, createGetDeviceServiceLoaderMockFail(), true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			op := NewDeviceServiceLoadAll(tt.cfg, tt.dbMock, logger.MockLogger{})
+			_, err := op.Execute()
+			if err != nil && !tt.expectError {
+				t.Error(err)
+				return
+			}
+			if err == nil && tt.expectError {
+				t.Errorf("error was expected, none occurred")
+				return
+			}
```

Comment:
Should we be checking for more than just the operator's return code? i.e., checking the number of services returned, checking an ID for sanity?

Reply:
In this case, I personally don't think so because everything around persistence is mocked. I have an interface that I'm glomming some logic onto through the mockery extensions, and there really isn't any kind of calculation or complex operation going on. My approach has been to target a specific execution flow in the code and then set up a mock to exactly satisfy that flow -- like if I need a particular exception to be returned. If there were some result obtained from passing in some params and then operating on them, then yes I would make that `expectedValue` part of my test criteria. This is sort of the approach taken [here](https://github.com/edgexfoundry/edgex-go/pull/1508/files/50c35001a25b6e50ca63384178fdf3e7cbce7f21#diff-d49e364aa4cf09204647193cd6b5a9ef) in testing the router. I've got a particular status I'm expecting from the operator, and the dbClient mock is setup in combination withe Configuration to ensure that condition is returned.

Reply:
I can understand your point, but the fact that a `return nil, nil` would satisfy the positive cases still kind of bothers me.  I think at the bare minimum I'd prefer checking to make sure the slice returned is the expected length, but if you don't think it's necessary I understand.

Reply:
It seems like there's nontrivial code happening between the mocks and the returns. We should be doing SOME sort of checking on the return value, whether that's checking if it's a nil slice in error cases or has an expected length in positive cases to ensure that inadvertent code changes don't slip through.

Reply:
@jdharms and @brandonforster - OK you convinced me.
BTW -- should we not have a test case to cover no device services registered without it being an error. That could happen at the beginning of time. Am I missing something?